syntax = "proto3";

package iterators;

message TriplePattern {
  string subject = 1;
  string predicate = 2;
  string object = 3;
  string graph = 4;
}

message SolutionMapping {
  map<string, string> bindings = 1;
}

message SavedScanIterator {
  TriplePattern pattern = 1;
  map<string, string> muc = 2;
  map<string, string> mu = 3;
  string last_read = 4;
  string timestamp = 5;
  int64 produced = 6;
  int64 cumulative_produced = 7;
  int64 cumulative_cardinality = 8;
  int64 stages = 11;
  float coverage = 12;
  float cost = 13;
  int64 cardinality = 14;
}

message SavedValuesIterator {
  repeated SolutionMapping values = 1;
  int64 next_value = 2;
  map<string, string> muc = 3;
  int64 produced = 4;
  float coverage = 5;
  float cost = 6;
}

message SavedProjectionIterator {
  repeated string values = 1;
  oneof source {
    SavedScanIterator scan_source = 2;
    SavedIndexJoinIterator join_source = 3;
    SavedBagUnionIterator union_source = 4;
    SavedFilterIterator filter_source = 5;
    SavedValuesIterator values_source = 6;
  }
  float coverage = 7;
  float cost = 8;
}

message SavedIndexJoinIterator {
  oneof left {
    SavedScanIterator scan_left = 1;
    SavedProjectionIterator proj_left = 2;
    SavedBagUnionIterator union_left = 3;
    SavedIndexJoinIterator join_left = 4;
    SavedFilterIterator filter_left = 5;
    SavedValuesIterator values_left = 6;
  }
  oneof right {
    SavedScanIterator scan_right = 7;
    SavedProjectionIterator proj_right = 8;
    SavedBagUnionIterator union_right = 9;
    SavedIndexJoinIterator join_right = 10;
    SavedFilterIterator filter_right = 11;
    SavedValuesIterator values_right = 12;
  }
  map<string, string> muc = 13;
  float coverage = 14;
  float cost = 15;
}

message SavedBagUnionIterator {
  oneof left {
    SavedScanIterator scan_left = 1;
    SavedProjectionIterator proj_left = 2;
    SavedBagUnionIterator union_left = 3;
    SavedIndexJoinIterator join_left = 4;
    SavedFilterIterator filter_left = 5;
    SavedValuesIterator values_left = 6;
  }
  oneof right {
    SavedScanIterator scan_right = 7;
    SavedProjectionIterator proj_right = 8;
    SavedBagUnionIterator union_right = 9;
    SavedIndexJoinIterator join_right = 10;
    SavedFilterIterator filter_right = 11;
    SavedValuesIterator values_right = 12;
  }
  float coverage = 13;
  float cost = 14;
}

message SavedFilterIterator {
  oneof source {
    SavedScanIterator scan_source = 1;
    SavedProjectionIterator proj_source = 2;
    SavedFilterIterator filter_source = 3;
    SavedIndexJoinIterator join_source = 4;
    SavedValuesIterator values_right = 5;
  }
  string expression = 6;
  repeated string variables = 7;
  map<string, string> mu = 8;
  int64 consumed = 9;
  int64 produced = 10;
  float coverage = 11;
  float cost = 12;
}

message SavedInsertData {
  map<string, uint64> nb_inserted = 1;
}

message SavedDeleteData {
  map<string, uint64> nb_inserted = 1;
}

message RootTree {
  oneof source {
    SavedScanIterator scan_source = 1;
    SavedProjectionIterator proj_source = 2;
    SavedIndexJoinIterator join_source = 3;
    SavedBagUnionIterator union_source = 4;
    SavedFilterIterator filter_source = 5;
    SavedInsertData insert_source = 6;
    SavedDeleteData delete_source = 7;
    SavedValuesIterator values_right = 8;
  }
}
